File: user_service.py
import mysql.connector
import subprocess
import os
from flask import Flask, request, render_template_string

app = Flask(__name__)

# Hardkodovani credentials
DB_CONFIG = {
    'host': 'localhost',
    'user': 'root',
    'password': 'password123',  # Hardkodovana lozinka
    'database': 'testdb'
}

# SQL Injection - direktna konkatenacija korisničkog inputa
def get_user_by_username(username):
    query = f"SELECT * FROM users WHERE username = '{username}'"
    
    conn = mysql.connector.connect(**DB_CONFIG)
    cursor = conn.cursor()
    cursor.execute(query)  # SQL injection!
    result = cursor.fetchone()
    cursor.close()
    conn.close()
    return result

# Još jedna SQL injection greška
def delete_user(user_id):
    sql = f"DELETE FROM users WHERE id = {user_id}"
    conn = mysql.connector.connect(**DB_CONFIG)
    cursor = conn.cursor()
    cursor.execute(sql)  # SQL injection!
    conn.commit()
    cursor.close()
    conn.close()

# XSS vulnerability - direktan output bez sanitizacije
@app.route('/comment', methods=['POST'])
def add_comment():
    comment = request.form.get('comment')
    template = f"<div>{comment}</div>"  # XSS!
    return render_template_string(template)

# Command injection
def execute_command(user_input):
    subprocess.call(['ping', user_input])  # Command injection!
    # Ili još gore:
    os.system(f"ping {user_input}")  # Command injection!

# Hardkodovani secret key
SECRET_KEY = "mySecretKey123"  # Hardkodovani secret

# Loša praksa - slanje lozinke u plain text-u
def login(username, password):
    # Nema hash-ovanja lozinke
    query = f"SELECT * FROM users WHERE username = '{username}' AND password = '{password}'"
    # SQL injection + plain text password!
    conn = mysql.connector.connect(**DB_CONFIG)
    cursor = conn.cursor()
    cursor.execute(query)
    result = cursor.fetchone()
    return result

# Nema validacije inputa
def process_payment(amount, card_number):
    # Nema validacije
    payment_data = {
        'amount': amount,
        'card_number': card_number
    }
    # Slanje payment podataka...

# Loša praksa - eval sa korisničkim inputom
def calculate_expression(expression):
    result = eval(expression)  # Opasno! Eval sa korisničkim inputom
    return result

# Loša praksa - pickle sa nevalidiranim podacima
import pickle
def load_user_data(data):
    user = pickle.loads(data)  # Opasno! Pickle može izvršiti kod
    return user

if __name__ == '__main__':
    app.run(debug=True)  # Debug mode u produkciji - loša praksa

File: userService.js
const mysql = require('mysql');

// Hardkodovani credentials u kodu
const connection = mysql.createConnection({
    host: 'localhost',
    user: 'root',
    password: 'password123', // Hardkodovana lozinka
    database: 'testdb'
});

// SQL Injection - direktna konkatenacija korisničkog inputa
function getUserByUsername(username) {
    const query = `SELECT * FROM users WHERE username = '${username}'`;
    
    connection.query(query, (error, results) => {
        if (error) {
            console.error('Error:', error); // Loša praksa - logovanje detalja greške
            return;
        }
        return results;
    });
}

// Još jedna SQL injection greška
function deleteUser(userId) {
    const sql = `DELETE FROM users WHERE id = ${userId}`;
    connection.query(sql, (err, result) => {
        if (err) throw err;
        console.log('User deleted');
    });
}

// XSS vulnerability - direktan output bez sanitizacije
function displayComment(comment) {
    document.getElementById('comments').innerHTML = comment; // XSS!
}

// Command injection mogućnost
function executeSystemCommand(command) {
    const { exec } = require('child_process');
    exec('ls -la ' + command, (error, stdout, stderr) => { // Command injection!
        if (error) {
            console.error(`exec error: ${error}`);
            return;
        }
        console.log(stdout);
    });
}

// Hardkodovani JWT secret
const JWT_SECRET = 'mySecretKey123'; // Hardkodovani secret

// Loša praksa - slanje lozinke u plain text-u
function login(username, password) {
    fetch('/api/login', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            username: username,
            password: password // Loša praksa - trebalo bi hash-ovati
        })
    });
}

// Nema validacije inputa
function processPayment(amount, cardNumber) {
    // Nema validacije da li je amount pozitivan broj
    // Nema validacije formata kartice
    const payment = {
        amount: amount,
        cardNumber: cardNumber
    };
    // Slanje payment podataka...
}

module.exports = {
    getUserByUsername,
    deleteUser,
    displayComment,
    executeSystemCommand,
    login,
    processPayment
};
